[IR][ModRef] Introduce `errno` memory location

antoniofrighetto opened the pull request:
Model C/C++ `errno` macro by adding a corresponding `errno` memory location kind to the IR. Preliminary work to separate `errno` writes from other memory accesses, to the benefit of alias analyses and optimization correctness.

Previous discussion: https://discourse.llvm.org/t/rfc-modelling-errno-memory-effects/82972.

antoniofrighetto commented:
Kind ping for correct direction.

efriedma-quic commented on llvm/test/Transforms/FunctionAttrs/writeonly.ll:
```diff
@@ -82,7 +82,7 @@ define void @test_store(ptr %p) {
 @G = external global ptr
 define i8 @test_store_capture(ptr %p) {
-; FNATTRS: Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(readwrite, argmem: read, inaccessiblemem: none)
+; FNATTRS: Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(readwrite, argmem: read, inaccessiblemem: none, errnomem: none)
```
How are we inferring `errnomem: none` here?

antoniofrighetto replied:
While walking the call-graph for deducing attrs, in `checkFunctionMemoryAccess` we initialize a ME object that doesn't access/modify memory and gradually refine it (currently says ErrnoMem: NoModRef). This ME is intersected with the original ME based on AA results (which says ErrnoMem: ModRef), so we get NoModRef, as per the meet of the lattice. We miss logic to infer if the function clobbers errno, but I think it should be fine as the PR only introduces the location (it might be better to conservatively always say ModRef, but the location is currently unused, so it should be okay to address this in an upcoming PR?). Rebased to main too.

efriedma-quic replied:
I'd prefer to get this right from the start, to avoid any future confusion.

efriedma-quic replied:
This is still wrong in the latest patch?

antoniofrighetto replied:
Hmm, the pointer is never accessed with an integer size, it should never alias errno, so should be correct to infer NoModRef for `test_store_capture`? Though now I think we might still miss something as follows while inferring attributes (and then let AA conclude if it aliases errno or not)?
```c++
if (Loc->Size == LocationSize::precise(sizeof(int)))
  ME |= MemoryEffects::errnoMemOnly(MR);
```

antoniofrighetto replied:
Mind checking that the last fixup commit is more aligned with what you thought? Can drop the check in addLocAccess if deemed unneeded.

antoniofrighetto replied:
Hmm, need to take a better look at this.

efriedma-quic reviewed (commented):
This seems like roughly what I expected from the proposal.

nikic commented on llvm/include/llvm/Analysis/MemoryLocation.h:
```diff
@@ -194,6 +194,10 @@ class LocationSize {
   }
 
+  bool operator<=(const LocationSize &Other) const {
+    return Value <= Other.Value;
```
LocationSize has a whole bunch of special values and bits that make this incorrect. You need to work on getValue() for the cases where it is valid.

antoniofrighetto replied:
Not really sure why I didn't use getValue in the first place.

nikic commented on llvm/include/llvm/AsmParser/LLToken.h:
```diff
@@ -201,11 +201,16 @@ enum Kind {
+  kw_inaccessiblemem_or_errnomemonly,
+  kw_argmem_or_errnomemonly,
+  kw_inaccessiblemem_or_argmem_or_errnomemonly,
+  kw_errnomemonly,
```
All of these shouldn't be there, they just exist to upgrade old IR.

nikic commented on llvm/include/llvm/Bitcode/LLVMBitCodes.h:
```diff
@@ -789,6 +789,10 @@ enum AttributeKindCodes {
+  ATTR_KIND_ERRNOMEMONLY = 103,
+  ATTR_KIND_INACCESSIBLEMEM_OR_ERRNOMEMONLY = 104,
+  ATTR_KIND_ARGMEM_OR_ERRNOMEMONLY = 105,
+  ATTR_KIND_INACCESSIBLEMEM_OR_ARGMEM_OR_ERRNOMEMONLY = 106,
```
Same here, these should not exist.

nikic commented on llvm/include/llvm/IR/Function.h:
```diff
@@ -587,11 +587,31 @@ class LLVM_ABI Function : public GlobalObject, public ilist_node<Function> {
+  ///  either inaccessible from the IR, pointed to by its arguments, or errno
+  ///  memory.
+  bool onlyAccessesInaccessibleMemOrArgMemOrErrnoMem() const;
+  void setOnlyAccessesInaccessibleMemOrArgMemOrErrnoMem();
```
Do not add these combinatorial APIs, unless they are widely useful.

antoniofrighetto replied:
I think there are a handful of libc routines which would benefit from these APIs.

antoniofrighetto replied:
That makes sense!

nikic commented on llvm/include/llvm/IR/InstrTypes.h:
```diff
@@ -1915,11 +1915,31 @@ class CallBase : public Instruction {
+  ///  either inaccessible from the IR, pointed to by its arguments, or errno
+  ///  memory.
+  bool onlyAccessesInaccessibleMemOrArgMemOrErrnoMem() const;
+  void setOnlyAccessesInaccessibleMemOrArgMemOrErrnoMem();
```
Same.

nikic commented on llvm/lib/Bitcode/Reader/BitcodeReader.cpp:
```diff
@@ -2293,9 +2318,21 @@ static bool upgradeOldMemoryAttribute(MemoryEffects &ME, uint64_t EncodedKind) {
+    return true;
+  case bitc::ATTR_KIND_INACCESSIBLEMEM_OR_ARGMEM_OR_ERRNOMEMONLY:
+    ME &= MemoryEffects::inaccessibleOrArgOrErrnoMemOnly();
+    return true;
```
As said, you don't need all of this upgrade code for attributes that never existed ... but you **do** need code to upgrade old MemoryEffects, where you need to copy Other to Errno.

A way to do that would be to add a version number to the high byte of the MemoryEffects bitcode encoding.

antoniofrighetto replied:
Thank you for clarifying this, I think it's a bit clearer now. I still have a doubt though: in `decodeLLVMAttributesForBitcode`, I believe I should expect to have something as follows:
```cpp
uint64_t Attrs = ((EncodedAttrs & (0xfffffULL << 32)) >> 11) | (EncodedAttrs & 0xffff);
uint8_t Version = (EncodedAttrs >> 56) & 0xFF;

if (AttrIdx == AttributeList::FunctionIndex) {
    // Upgrade old memory attributes.
    // ...
    // If version < 1 and attribute for Other is set, copy Other ModRef to Errno.
    if (Version < 1 && (Attrs & (7ULL << 39))
        ME &= MemoryEffects::errnoMemOnly(ME.getModRef(IRMemLocation::Other));

    if (ME != MemoryEffects::unknown())
      B.addMemoryAttr(ME);
}
```
However, we currently miss a mask for attribute Other to check if it was set, so this should be introduced too?

> A way to do that would be to add a version number to the high byte of the MemoryEffects bitcode encoding.

As per doc-comment in `decodeLLVMAttributesForBitcode`, I was expecting to see `encodeLLVMAttributesForBitcode` in the Writer, where to add a version number, but seems to have been dropped a while ago. Am I understanding correctly that in `writeAttributeTable` we should somehow still have something as follows?
```cpp
  const uint8_t Version = 1;
  uint64_t EncodedAttrs = (Version << 56) | ME.getAsInt();
  Record.emplace_back(EncodedAttrs);
```
This looks a little bit confusing though, as we don't seem to be _directly_ manipulating MemoryEffects encoding anymore (as it used to be the case).

nikic replied:
decodeLLVMAttributesForBitcode is not relevant for this kind of attribute. The place to perform the upgrade would be here: https://github.com/llvm/llvm-project/blob/83af335ea47b50037beb46e5d6fb04be89f3b207/llvm/lib/Bitcode/Reader/BitcodeReader.cpp#L2401-L2402

And the version can be added on the encoding side here: https://github.com/llvm/llvm-project/blob/83af335ea47b50037beb46e5d6fb04be89f3b207/llvm/lib/Bitcode/Writer/BitcodeWriter.cpp#L981 This would be handling just for MemoryEffects, not generically for all attributes.

antoniofrighetto replied:
Thank you, I think it should look like towards being more on track now.

nikic commented on llvm/lib/Transforms/IPO/FunctionAttrs.cpp:
```diff
@@ -126,8 +126,21 @@ static void addLocAccess(MemoryEffects &ME, const MemoryLocation &Loc,
+      ME |= MemoryEffects::errnoMemOnly(MR);
+      return;
+    }
+  }
```
Let's please not introduce any special inference logic for errno in this patch, we can do this in a followup.

nikic replied:
Note that the bit you actually *want* to modify the the code a few lines below that uses IRMemLocation::Other. That one should not include Errno as well.

antoniofrighetto replied:
>  That one should not include Errno as well

You meant that we should include the memory effects of Errno, as Other doesn't affect Errno, correct?

nikic replied:
Yes

nikic commented on llvm/lib/Transforms/IPO/FunctionAttrs.cpp:
```diff
@@ -126,8 +126,21 @@ static void addLocAccess(MemoryEffects &ME, const MemoryLocation &Loc,
   if (isa<Argument>(UO)) {
     ME |= MemoryEffects::argMemOnly(MR);
+    if (Loc.Size <= LocationSize::precise(sizeof(int)))
+      ME |= MemoryEffects::errnoMemOnly(MR);
```
This doesn't look right. Even if the argument might alias errno, it's still argmem from the perspective of this function. The caller has to deal with what the argument actually points to.

antoniofrighetto replied:
Dropped the check for now. I'm slightly unsure if it'd still make sense to keep it while examining formal parameters within the call-site (hence, from caller perspective)? Thinking more on checking the memory location size, say, e.g., a memory location possibly refers to a pointer to pointer: this case would be incorrectly excluded from the above check, correct?

antoniofrighetto committed:
[IR][ModRef] Introduce `errno` memory location

Model C/C++ `errno` macro by adding a corresponding `errno`
memory location kind to the IR. Preliminary work to separate
`errno` writes from other memory accesses, to the benefit of
alias analyses and optimization correctness.

Previous discussion: https://discourse.llvm.org/t/rfc-modelling-errno-memory-effects/82972.

nikic reviewed (commented):
This needs a LangRef update.

nikic commented on llvm/lib/Bitcode/Writer/BitcodeWriter.cpp:
```diff
@@ -977,8 +977,15 @@ void ModuleBitcodeWriter::writeAttributeGroupTable() {
+        Record.push_back(getAttrKindEncoding(Kind));
+        if (Kind == Attribute::Memory) {
+          // Version field for upgrading old memory effects.
+          const uint64_t Version = 1 << 0;
```
```suggestion
          const uint64_t Version = 1;
```

nikic commented on llvm/lib/Bitcode/Reader/BitcodeReader.cpp:
```diff
@@ -2398,9 +2397,16 @@ Error BitcodeReader::parseAttributeGroupBlock() {
+                                                        0x00FFFFFFFFFFFFFFULL);
+            if (Version < 1)
+              ME |= MemoryEffects::errnoMemOnly(
+                  ME.getModRef(IRMemLocation::Other));
```
This is not going to do the right thing, because the old Other will be located at the new ErrnoMem, so what you'd actually have to do is copy from errno to other. It might be better to full reconstruct the MemoryEffects without going through createFromIntValue for old versions.

In any case, this needs a test. You need to create a bitcode file with the old llvm-as that has something like `memory(write, argmem: read)` and then check that llvm-dis still produces the same result (instead of `memory(errnomem: write, argmem: read)` which I think is what happens without upgrade).

antoniofrighetto replied:
Updated, thanks, should have introduced a test earlier. Thus, IIUC, for older versions, errno should be part of default memory.

nikic commented on llvm/test/Transforms/SCCP/ipscp-drop-argmemonly.ll:
```diff
@@ -11,10 +11,10 @@
 ;.
 define internal void @ptrarg.1(ptr %arg, i32 %val) argmemonly nounwind {
-; CHECK: Function Attrs: nounwind memory(readwrite, inaccessiblemem: none)
+; CHECK: Function Attrs: nounwind memory(readwrite, inaccessiblemem: none, errnomem: none)
```
I wonder if we should be setting errnomem as well in https://github.com/llvm/llvm-project/blob/64735ad63975c1126f84dbf7921ce4341dfa2419/llvm/lib/Transforms/IPO/SCCP.cpp#L194-L195. The propagated global might be errno.

antoniofrighetto replied:
That's true, added this as well.

jdoerfert commented:
There is no more deduction in this commit, right? And no propagation tests either, correct?

antoniofrighetto commented:
> There is no more deduction in this commit, right?

Right, will be introduced in a followup PR.

nikic commented on llvm/include/llvm/Support/ModRef.h:
```diff
@@ -139,6 +141,16 @@ template <typename LocationEnum> class MemoryEffectsBase {
+  }
+
+  /// Create MemoryEffectsBase that can only access default memory.
+  static MemoryEffectsBase defaultMemOnly(ModRefInfo MR = ModRefInfo::ModRef) {
```
```suggestion
  static MemoryEffectsBase otherMemOnly(ModRefInfo MR = ModRefInfo::ModRef) {
```
To keep terminology consistent?

nikic commented on llvm/lib/Bitcode/Reader/BitcodeReader.cpp:
```diff
@@ -2398,9 +2397,25 @@ Error BitcodeReader::parseAttributeGroupBlock() {
+                   MemoryEffects::argMemOnly(
+                       ME.getModRef(IRMemLocation::ArgMem)) |
+                   MemoryEffects::errnoMemOnly(OldOtherMR) |
+                   MemoryEffects::defaultMemOnly(OldOtherMR);
```
What I had in mind here was more along the lines of:
```
ModRefInfo ArgMem = ModRefInfo((EncodedME >> 0) & 3);
ModRefInfo InaccessibleMem = ModRefInfo((EncodedME >> 2) & 3);
ModRefInfo OtherMem = ModRefInfo((EncodedME >> 4) & 3);
ME = MemoryEffects::inaccessibleMemOnly(InaccessibleMem) |
                   MemoryEffects::argMemOnly(ArgMem) |
                   MemoryEffects::errnoMemOnly(OtherMem) |
                   MemoryEffects::defaultMemOnly(OtherMem);
```
Basically to entirely avoid instantiating MemoryEffects with the old value.

antoniofrighetto replied:
Right, indeed looks much better this way.

nikic commented on llvm/include/llvm/Support/ModRef.h:
```diff
@@ -139,6 +141,16 @@ template <typename LocationEnum> class MemoryEffectsBase {
+    return MemoryEffectsBase(Location::ErrnoMem, MR);
+  }
+
+  /// Create MemoryEffectsBase that can only access default memory.
```
```suggestion
  /// Create MemoryEffectsBase that can only access other memory.
```

nikic commented on llvm/lib/Bitcode/Reader/BitcodeReader.cpp:
```diff
@@ -2398,9 +2397,28 @@ Error BitcodeReader::parseAttributeGroupBlock() {
+          else if (Kind == Attribute::Memory) {
+            uint64_t EncodedME = Record[++i];
+            const uint8_t Version = (EncodedME >> 56);
+            if (Version < 1) {
```
```suggestion
            if (Version == 0) {
```

nikic reviewed (approved):
LGTM
